### ISSUE TITLE
synthesis quality

### ISSUE THREAD

**Issue by KiaraGrouwstra** [tycho01/synthesis#10]
### experiments

- [x] confirm accuracy/learning
- [x] [early stopping](https://github.com/tycho01/uvadlc_practicals_2019/blob/master/assignment_1/code/train_convnet_pytorch.py#L126-L134)
- [x] [lr decay](https://github.com/tycho01/uvadlc_practicals_2019/blob/master/assignment_1/code/train_convnet_pytorch.py#L136-L139)
- hyper-params
  - hyperparameter optimization:
    - [x] implement grid search, including distinguishing validation vs training sets
    - [x] implement evolutionary search (see [1](https://github.com/tycho01/uvadlc_practicals_2019/blob/master/assignment_1/code/ea.py), [2](https://github.com/tycho01/uvadlc_practicals_2019/blob/master/assignment_1/code/train_mlp_pytorch.py#L203-L222))
  - ~~[ ] consider `LearnedInitialization` for LSTMs~~
- [x] [let a learner pick which hole to fill](https://github.com/tycho01/synthesis/blob/cc42add9a47417c38eb88ac92b298297e3a8c858/library/Ast.hs#L125)
- [x] add explicit type signature for this i/o so programs won't fail type check over uninferred types (which can be a problem for e.g. `Either`, or for `Maybe` in case of `Nothing`)
- [x] try incorporating expansion embeddings in node embeddings thru either pass?
- [x] can I meaningfully encode anything more than just 'expression' in symbols?
  - [x] hole status?
  - [x] fn arities, as seemingly done in NSPS?
  - [x] block (before variants)?
  - [x] types?
